enhancements/rhcos/extensions.md

```diff
@@ -0,0 +1,96 @@
+
+We propose continuing the trail blazed with the RT kernel by adding additional RPMs to `machine-os-content` that aren't part of the base OS "image", but can be installed later.  This is how `kernel-rt` works today; we added a `kernel-rt.rpm` that is committed to the container.  In the future though, we may instead ship a separate `machine-os-extensions` container, or something else.  The only API stable piece here is the `MachineConfig` field (same as for the `kernelType: realtime`).
+
+An important way to think about this is that we can still ascribe a *single version number* to this set of content - all content is still versioned with and tested with the OS (and hence the main OpenShift 4 release image).
+
+### Goals
+
+- Add support for cluster operators (notably the MCO) to deploy non-containerized software that is maintained alongside OpenShift
+- Avoid forcing the authors of software like `usbguard` to maintain two ways to ship their software (as RPM and as a container)
+
+### Non-Goals
+
+- Direct support for installing RPMs *not* from the release image
+
+## Proposal
+
+1.  RHCOS build system is updated to inject `usbguard` (and other software) into `machine-os-content` as an RPM
+2.  MachineConfig gains support for `extensions` that are layered on in the same way it applies OS updates and manages `kernel-rt`
+
+### Implementation Details/Notes/Constraints
+
+Note this will also likely enable us to move `open-vm-tools` out of the host and have the MCO install it on VSphere environments for example.
+
+### Risks and Mitigations
+
+#### 3rd party RPMs
+
+This will blaze a trail that will make it easier to install 3rd party RPMs, which is much more of a risk in terms of compatibility and for upgrades.
+
+#### Non-Kubernetes native management
+
+How one *ships* software has a huge impact on how one *manages* the software.  Shipping software as an RPM implies nowadays it usually runs as a `systemd` unit, parses config files in `/etc`, logs to the journal etc.
+
+When one ships software as a container for Kubernetes, one can use `oc/kubectl` to inspect its status, it might offer custom resource definitions, etc.  And we have made investment in ensuring a "Kubernetes native" interface to important components.
+
+Writing containers should be preferred - but, we cannot do that all at once, for all software that is relevant for host execution.
+
+### Upgrades
+
+The MCO will update the OS and extensions as one unit.  All shipped extensions will have been tested with the same version of the host.
+
+## Alternatives
+
+Multiple OS builds: Becomes a combinatorial nightmare.
+
+Do nothing: Probably RHCOS will continue to grow and we will have to look embarrassed when someone asks us why `usbguard` is installed in their AWS cluster where the VMs don't have USB.
+
+Force `usbguard` authors to containerize: Not a technical problem exactly but...it is *really* hard to have two ways to ship software; see above.
```

Comment:
Isn't this the point of the container revolution? You move away from the old methods of packaging and delivering software, to containers (so that its the standard)? - insert xkdc meme for standards. 

It seems to be that were trying to solve a problem for people because they are obstinate (yes change is hard, and I am empathetic to that, but how / why is that OpenShift's issue).

Reply:
Really what we're trying to do is not ship all the things in base to meet all the requirements. To do would increase the size of the boot disks. Instead, we're trying to push off things which are needed in specific scenarios (Example: `usbguard` in environments with exposed USB who care about security) rather than "put it in base".